File: 0078_subsets/subsets.py
from typing import List

def subsets(nums: List[int]) -> List[List[int]]:
    """
    Time: O(n * 2**n)
    Space: O(n * 2**n)
    """
    res = []
    dfs(nums, 0, [], res)
    return res

def dfs(nums: List[int], index: int, path: List[int], res: List[List[int]]):
    res.append(path)
    for i in range(index, len(nums)):
        dfs(nums, i+1, path+[nums[i]], res)

File: 0078_subsets/subsets.go
package main

// Time: O(n * 2**n)
// Space: O(n * 2**n)
func subsets(nums []int) [][]int {
	res := [][]int{}
	dfs(nums, 0, []int{}, &res)
	return res
}

func dfs(nums []int, index int, path []int, res *[][]int) {
	*res = append(*res, append([]int{}, path...))
	for i := index; i < len(nums); i++ {
		path := append(path, nums[i])
		dfs(nums, i+1, path, res)
		path = path[:len(path)-1]
	}
}

func main() {

}
